src/library/dao/trackdao.cpp

```diff
@@ -2137,3 +2189,36 @@ TrackFile TrackDAO::relocateCachedTrack(
         return TrackFile(trackLocation);
     }
 }
+
+bool TrackDAO::updatePlayCounterFromHistoryPlaylists(
+        const QSet<TrackId> trackIds) const {
+    FwdSqlQuery query(
+            m_database,
+            QStringLiteral(
+                    "UPDATE library SET "
+                    "timesplayed=q.timesplayed,"
+                    "last_played_at=q.last_played_at "
+                    "FROM("
+                    "SELECT "
+                    "PlaylistTracks.track_id as id,"
+                    "COUNT(PlaylistTracks.track_id) as timesplayed,"
+                    "MAX(PlaylistTracks.pl_datetime_added) as last_played_at "
+                    "FROM PlaylistTracks "
+                    "JOIN Playlists ON "
+                    "PlaylistTracks.playlist_id=Playlists.id "
+                    "WHERE Playlists.hidden=%2 "
+                    "GROUP BY PlaylistTracks.track_id"
+                    ") q "
+                    "WHERE library.id=q.id "
+                    "AND library.id IN (%1)")
+                    .arg(joinTrackIdList(trackIds),
+                            QString::number(PlaylistDAO::PLHT_SET_LOG)));
+    VERIFY_OR_DEBUG_ASSERT(!query.hasError()) {
+        return false;
+    }
```

Comment:
I think we need also consider the played flag.

Maybe we can dispose it internally by comparing the current play count with the value the session starts. That sounds less redundant.

I did not have a look, but we have also a feature to join two set logs, the played flag needs to be considered than as well.

Reply:
The *played* flag should not be a property of the (persistent) track object. The set of tracks that are considered as (manually) played during a session should be managed separately. The complicated and inconsistent logic in the `PlayCounter` is already screaming out loud. I refuse to add more *magic* to it that fits some use cases but induces unexpected behavior for others. If the user decides to rewrite history they could be bothered with updating the *played* flag manually. It will be reset anyway at some time.

For my aoide view I don't have a *played* flag and already use the logic you proposed for an implicit *played* flag. This could be overridden with a session *played* flag, once Mixxx supports this concept.

Reply:
I agree that the played flag is a slightly different idea and should remain separate.  (Also note that Played status is maintained across crashes so the user doesn't lose state :))

Reply:
Mangling the *played flag* as a column of the Track entity table was a bad idea. Instead the set of played tracks must be stored separately by introducing the concept of a session. Out of scope of this PR.